File: TwoPointer/53.maximum-subarray.go
package leetcode

// @leet start
func maxSubArray(nums []int) int {
	cum, res := nums[0], nums[0]
	for i, v := range nums {
		if i == 0 {
			continue
		}
		if cum > 0 {
			cum += v
		} else {
			cum = v
		}
		res = max(res, cum)

	}
	return res
}

// @leet end

File: TwoPointer/53.maximum-subarray.py
from typing import List


# @leet start
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        best = total = nums[0]
        for i in range(1, len(nums)):
            item = nums[i]
            # if total + item > total:
            # if item + total > item:  # NOTE: chose total!!
            if total > 0:  # NOTE: chose total!!
                total += item
            else:
                total = item
            best = max(best, total)
        return best


# @leet end

